src/libstore/build.cc

```diff
@@ -3689,162 +3835,314 @@ void DerivationGoal::registerOutputs()
-                    BuildError("hash mismatch in fixed-output derivation '%s':\n  wanted: %s\n  got:    %s",
-                        worker.store.printStorePath(dest),
-                        h.to_string(SRI, true),
-                        h2.to_string(SRI, true)));
-              } else {
-                  // matched the fixed hash, so no move needed.
-                  needsMove = false;
-              }
+            rewriteOutput();
+            /* FIXME optimize and deduplicate with addToStore */
+            std::string oldHashPart { scratchPath.hashPart() };
+            HashModuloSink caSink { outputHash.hashType, oldHashPart };
+            switch (outputHash.method) {
+            case FileIngestionMethod::Recursive:
+                dumpPath(actualPath, caSink);
+                break;
+            case FileIngestionMethod::Flat:
+                readFile(actualPath, caSink);
+                break;
+            }
+            auto got = caSink.finish().first;
+            auto refs = rewriteRefs();
+            ValidPathInfo newInfo0 {
+                worker.store.makeFixedOutputPath(
+                    outputHash.method,
+                    got,
+                    outputPathName(drv->name, outputName),
+                    refs.second,
+                    refs.first),
+            };
+            newInfo0.ca = FixedOutputHash {
+                .method = outputHash.method,
+                .hash = got,
+            };
+            newInfo0.references = refs.second;
+            if (refs.first)
+                newInfo0.references.insert(newInfo0.path);
+            {
+                HashModuloSink narSink { htSHA256, oldHashPart };
+                dumpPath(actualPath, narSink);
+                auto narHashAndSize = narSink.finish();
+                newInfo0.narHash = narHashAndSize.first;
+                newInfo0.narSize = narHashAndSize.second;
             }
 
-            if (needsMove) {
-                Path actualDest = worker.store.Store::toRealPath(dest);
+            /* Check wanted hash if output is fixed */
+            if (auto p = std::get_if<DerivationOutputCAFixed>(&output.output)) {
+                Hash & wanted = p->hash.hash;
+                if (wanted != got) {
+                    /* Throw an error after registering the path as
+                       valid. */
+                    worker.hashMismatch = true;
+                    delayedException = std::make_exception_ptr(
+                        BuildError("hash mismatch in fixed-output derivation '%s':\n  wanted: %s\n  got:    %s",
+                            worker.store.printStorePath(drvPath),
+                            wanted.to_string(SRI, true),
+                            got.to_string(SRI, true)));
+                }
+            }
 
-                if (worker.store.isValidPath(dest))
-                    std::rethrow_exception(delayedException);
+            assert(newInfo0.ca);
+            return newInfo0;
+        }}();
```

Comment:
Possibly rather subjective, but I'd find the control flow easier to follow if it was structured more like:

```cpp
auto newInfoFromCA = [&](const DerivationOutputCAFloating outputHash) -> ValidPathInfo {
// everything from L3999 to L4061
};
ValidPathInfo newInfo = [&]() std::visit(overloaded {
  [&](DerivationOutputInputAddressed outputP) { ... };
  [&](DerivationOutputCAFixed dof) { return newInfoFromCA(DerivationOutputCAFloating { ... }); };
  [&](DerivationOutputCAFloating dof) { return newInfoFromCA(DerivationOutputCAFloating dof; };
})
```

Reply:
Ah great idea! I agree this is better. And furthermore, the code that checks the wanted vs got can be moved out of the `newInfoFromCA` and into the `DerivationOutputCAFixed` case, avoiding yet another `std::get_if`.